feat: support returning and resuming partial results

ShamrockLee opened the pull request:
### Description

Add a parameter `keep_failed` to `Dials.list_all()`, specifying whether to return the paginated partial results when an HTTP request fails.

Add a parameter `resume_from` to `Dials.list_all()` to accept paginated partial results and resume the fetching.

Add a field `exception` to `PaginatedBaseModel` to record the exception that causes the request to fail.

If applied, users would be able to write something like:
```python
data = None
data = dials.h1d.list_all(<filter>, keep_failed=True, resume_from=data)
```
and resume by <kbd>Up</kbd><kbd>Enter</kbd> from the terminal.

### TODO

*   Update the documentation.
*   Run the tests.

ShamrockLee committed:
feat: support returning and resuming partial results

Add a parameter `keep_failed` to `Dials.list_all()`,
specifying whether to return the paginated partial results
when an HTTP request fails.

Add a parameter `resume_from` to `Dials.list_all()`
to accept paginated partial results and resume the fetching.

Add a field `exception` to `PaginatedBaseModel`
to record the exception that causes the request to fail.

If applied, users would be able to write something like:
```python
data = None
data = dials.h1d.list_all(<filter>, keep_failed=True, resume_from=data)
```
and resume by <Up><Enter> from the terminal.

ShamrockLee committed:
feat: support returning and resuming partial results

Add a parameter `keep_failed` to `Dials.list_all()`,
specifying whether to return the paginated partial results
when an HTTP request fails.

Add a parameter `resume_from` to `Dials.list_all()`
to accept paginated partial results and resume the fetching.

Add a field `exception` to `PaginatedBaseModel`
to record the exception that causes the request to fail.

If applied, users would be able to write something like:
```python
data = None
data = dials.h1d.list_all(<filter>, keep_failed=True, resume_from=data)
```
and resume by <Up><Enter> from the terminal.

gabrielmscampos commented:
I manage to test the resume feature (without the `Enter` interaction) adding these lines to the code:

```python
        while is_last_page is False:
            next_token = parse_qs(urlparse(next_string).query).get("next_token") if next_string else None
            curr_filters = self.filter_class(**filters.dict())
            curr_filters.next_token = next_token
            try:
                response = self.list(curr_filters)
                if total_pages == 10:
                    raise RuntimeError("Test")
            except Exception as e:
```

and updating the warning line to:

```python
                warn(
                    "HTTP request failed, returning partial results. Exception: " + "\n".join(str(e)),
                    RuntimeWarning,
                    stacklevel=2,
                )
```

Because `format_exception_only(e)` was generating another exception.

I tested with the following snippet:

 ```python
 from cmsdials.auth.client import AuthClient
from cmsdials.auth.bearer import Credentials
from cmsdials import Dials
from cmsdials.filters import LumisectionHistogram2DFilters

auth = AuthClient()
creds = Credentials.from_creds_file()
dials = Dials(creds, workspace="tracker")


data = None
data = dials.h2d.list_all(LumisectionHistogram2DFilters(me__regex="PXBarrel", ls_number=78, entries__gte=100), max_pages=59, keep_failed=True, resume_from=data)
print(len(data.results))  # 100

# After failing, run this again
data = dials.h2d.list_all(LumisectionHistogram2DFilters(me__regex="PXBarrel", ls_number=78, entries__gte=100), max_pages=59, keep_failed=True, resume_from=data)
print(len(data.results))  # 200
```

I liked the idea for more interactive environments. Do you plan to still work on it?

gabrielmscampos committed:
Merge branch 'resume' of github.com:ShamrockLee/dials-py into resume

gabrielmscampos committed:
fix: format_exception_only was raising another exception, replace it with format_exc

gabrielmscampos committed:
refactor: disable arbitrary_types_allowed in PaginatedBaseModel and store exc as string

- In pydantic v1, model.schema() fails when using arbitrary types. More
info: https://github.com/pydantic/pydantic/issues/4079
- Since exception as BaseException has no uses for now, we can safely
transform it to a string and also store the exc_type for future debug

gabrielmscampos committed:
feat: allow resuming even if no exception ocurred

- If we list an endpoint enforcing `max_pages`, we can return the last
`next` string and resume from there

gabrielmscampos committed:
refactor: update `setup_dials_object` in tests to optionally specify workspace

gabrielmscampos committed:
refactor: add test for resume feature

gabrielmscampos committed:
docs: update readme with resume instructions

gabrielmscampos commented:
I removed the `arbitrary_types_allowed` field from the `PaginatedBaseModel` since in Pydantic v1 this was causing `model.schema()` to crash since arbitrary types are not necessarily JSON serializable. Since `exception`, as far as I could see, was meant for debugging I replaced it with `exc_formatted` and `exc_type` so we potentially get almost the same debug capability without enabling `arbitrary_types_allowed`.

Also, I'm returning the `next` string in the final response (if no exceptions occur) so we can successfully resume from partial objects even if they haven't failed (added a test for that).

I'll merge it now, many thanks!
